ICEBERG-13343: fix race condition in JdbcCatalog

Claudenw opened the pull request:
Fix for ICEBERG-13343

Closes #13343

jbonofre commented on core/src/main/java/org/apache/iceberg/jdbc/JdbcUtil.java:
```diff
@@ -123,7 +123,7 @@ enum SchemaVersion {
           + " = ?";
   static final String V0_CREATE_CATALOG_SQL =
-      "CREATE TABLE "
+      "CREATE TABLE IF NOT EXISTS "
```
Afair, it has been discussed while ago.

As this syntax is not supported by all SQL backends, we control that via `JdbcUtil.INIT_CATALOG_TABLES_PROPERTY`.

You can see the login in the `initializeCatalogTables()` method.

jbonofre reviewed (commented):
Imho, this PR is not relevant related to previous discussions and logic used in the init method.

Claudenw commented:
Based on @jbonofre comments I have reworked the code to not use "IF NOT EXISTS" and instead to check if the table was created after the creation fails.  This case will occur during the race condition.

Added Apache Derby as a test DB as SqlLite does not handle multiple connections well.

Refactored some code to make it cleaner and easier to read.

kevinjqliu commented on core/src/test/java/org/apache/iceberg/jdbc/TestJdbcTableConcurrency.java:
```diff
@@ -149,4 +190,1015 @@ public synchronized void testConcurrentConnections() throws InterruptedException
+      }
+    } finally {
+      DriverManager.deregisterDriver(slowDriver);
+    }
```
nit: add a catalog operation here to verify that both catalog clients are initialized and ready

kevinjqliu commented on core/src/test/java/org/apache/iceberg/jdbc/TestJdbcTableConcurrency.java:
```diff
@@ -149,4 +190,1015 @@ public synchronized void testConcurrentConnections() throws InterruptedException
+    String testingDB = "jdbc:slow:derby:memory:testDb;create=true";
+    new org.apache.derby.jdbc.EmbeddedDriver();
+    properties.put(CatalogProperties.URI, testingDB);
+    SlowDriver slowDriver = new SlowDriver(testingDB);
```
Curious if we can test this without `SlowDriver`.

Can we simulate the same behavior by raising the parallelism? 
For example, `testConcurrentConnections` uses 7 threads

kevinjqliu reviewed (commented):
Thanks for adding this change to address the race condition during JDBC catalog initialization.


I think JB made a good point in the [other thread](https://github.com/apache/iceberg/issues/13343#issuecomment-2985448183),
```
We are suppose to create table at JDBC Catalog init, before the sink is "running".
```
This is a good workaround, and this PR eliminates the need for it.

I added a few nit comments. Please take a look

kevinjqliu reviewed (approved):
LGTM!

jbonofre commented:
Lgtm
I would just have split in 2 PRs:
One to show the issue (using derby) one to fix 

But only for me
